src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h

```diff
@@ -0,0 +1,293 @@
+    const std::size_t cbmc_loop_id,
+    const std::set<exprt> assigns,
+    const exprt invariant,
+    const exprt decreases,
+    const std::set<irep_idt> local,
+    const std::set<irep_idt> tracked,
+    const std::vector<std::size_t> inner_loops,
+    const std::vector<std::size_t> outer_loops,
+    const symbol_exprt write_set_var,
+    const symbol_exprt addr_of_write_set_var)
+    : loop_id(loop_id),
+      cbmc_loop_id(cbmc_loop_id),
+      assigns(assigns),
+      invariant(invariant),
+      decreases(decreases),
+      local(local),
+      tracked(tracked),
+      inner_loops(inner_loops),
+      outer_loops(outer_loops),
+      write_set_var(write_set_var),
+      addr_of_write_set_var(addr_of_write_set_var)
+  {
+  }
+
+  /// Prints a textual representation of the struct to \p out.
+  void output(std::ostream &out, namespacet &ns, const irep_idt &mode) const;
+
+  // Finds the instruction tagged as loop head and having the same loop
+  // identifier as this struct in the given program.
+  optionalt<goto_programt::instructiont::targett>
+  find_head(goto_programt &goto_program) const;
+
+  // Finds the last instruction tagged as loop latch hand having the same loop
+  // loop identifier as this struct in the given program.
+  optionalt<goto_programt::instructiont::targett>
+  find_latch(goto_programt &goto_program) const;
+
+  /// Loop identifier assigned by DFCC to this loop.
+  const std::size_t loop_id;
+
+  /// Loop identifier assigned to this loop by traditional CBMC loop numbering.
+  const std::size_t cbmc_loop_id;
+
+  /// Set of targets assigned by the loop, either user-provided or inferred.
+  const std::set<exprt> assigns;
+
+  /// Loop invariant expression.
+  const exprt invariant;
+
+  /// Decreases clause expression.
+  const exprt decreases;
+
+  /// Set of local identifiers locally DECL in loop instructions, excluding
+  /// identifiers declared in nested loops.
+  const std::set<irep_idt> local;
+
+  /// Subset of locals that must be tracked in the loop's write set.
+  /// A local must be tracked if it is dirty or might be assigned by one of
+  /// the inner loops of that loop.
+  const std::set<irep_idt> tracked;
+
+  /// Integer identifiers of inner loops of that loop.
+  const std::vector<std::size_t> inner_loops;
+
+  /// Integer identifier of the outer loop(s) if it exists.
+  const std::vector<std::size_t> outer_loops;
+
+  /// Symbol representing the stack-allocated write set object for this loop.
+  const symbol_exprt write_set_var;
+
+  /// Symbol representing pointer to the stack allocated write set object for
+  /// this loop. This is the one that must be passed as argument to write set
+  /// functions.
+  const symbol_exprt addr_of_write_set_var;
+};
+
+/// Computes natural loops, enforces normal form conditions, computes the
+/// nesting graph, tags each instruction of the function with the loop ID of the
+/// innermost loop and loop instruction type. Loop identifiers range from 0 to
```

Comment:
Oh wow, that's quite a lot of work in a single unit?! Is that the right thing to do? Also, I am saying "unit," because I don't actually know which method it is that does this - the comment is attached to the class. Is it the constructor? Constructors should only be doing that: construct an object.

Reply:
We want this to be a transaction so that there is now way to modify the goto program before the `dfcc_cfg_infot` instance has been created. So normalisation/tagging/mapping the loops etc has to be done all at once.

Reply:
But then it could be done in a function (perhaps not even a method, you might want a free function).

Reply:
How about putting 
> computes natural loops, enforces normal form conditions, computes the
> nesting graph, tags each instruction of the function with the loop ID of the
> innermost loop and loop instruction type
in a preprocess function. My question is where should we put such preprocess function.

1. calling the preprocess function from the constructor of `dfcc_cfg_infot`, which is basically the same as what we have now except for encapsulating all the preprocess into a function,
2. calling the prepossess function right before constructing `dfcc_cfg_infot` objects. But I am not sure if we will have a good way to ask other developers to call preprocess before the constructor. 

Which one sounds better for you?

Reply:
Couldn't you also construct a `dfcc_cfg_infot` object and then have the user of that object (the caller of the constructor) invoke a further method that does the heavy lifting? As an alternative I'd suggest:
1) Have a free function that's the actual interface (signature will be similar to the current constructor) and have
2) Move the class to the implementation (.cpp) file where that free function will do the above (i.e., construct an object and then invoke a method on it).

Reply:
I'm ok with solution 2.

Reply:
I'm keeping the initialisation logic in the constructor. Aren't constructors supposed to be establish the desired state as soon as the object is created ?